tests/python_test.py

```diff
@@ -54,16 +54,16 @@
 LOCKSPACE_OR_RESOURCE_NAMES = [
     # Bytes are supported with python 2 and 3.
     pytest.param(
-        b"\xd7\x90",
-        marks=pytest.mark.xfail(six.PY3, reason="python 3 should support only bytes")),
+        b"\xd7\x90"),
+        #marks=pytest.mark.xfail(six.PY3, reason="python 3 should support only bytes")),
     # Python 2 also supports str.
     pytest.param(
         "\xd7\x90",
         marks=pytest.mark.skipif(six.PY3, reason="python 3 supports only bytes")),
     # Python 2 also supports unicode with ascii content.
     pytest.param(
         u"ascii",
-        marks=pytest.mark.skipif(six.PY3, reason="python 3 supports only bytes")),
+        marks=pytest.mark.xfail(reason="python 2 does not support unicode string, python 3 only supports bytes")),
```

Comment:
This is a problem, since vdsm is using unicode names with ascii content now. This change will break vdsm using this code, or any other client that happen to use unicode strings. This is pretty common, since:

```
>>> import json
>>> json.loads(json.dumps({"name": "ascii"}))
{u'name': u'ascii'}
```

I think we will need to use "O&" format, and add a function that convert unicode
object to bytes on py2, and copy bytes object otherwise. Something like:

```
/* Return new reference via addr */
static int
convert_to_pybytes(PyObject *arg, void *addr)
{
#if PY_MAJOR_VERSION == 2
    if (PyUnicode_Check(arg)) {
        PyObject *bytes = PyUnicode_AsASCIIString(arg);
        if (bytes == NULL)
            return 0;

        *(PyObject **)addr = bytes;
        return 1;
    }
#endif

    if (PyBytes_Check(arg)) {
        Py_INCREF(arg);
        *(PyObject **)addr = arg;
        return 1;
    }
    
    set type error...
    return 0; 
}
```

Reply:
if we expect bytes in Py3, unicode/ascii strings would still break there; why using bytes for lockspace/resource names anyway?
This brings back to me earlier comment we that could be happy with using ascii strings for those only py3 string will support it. maybe we just need to make sure we only allow ascii strings in py3 for lockspace and resoruce name for interoperability between py2 and py3 vdsms and document this limitation in the user-guide when one day someone outside ovirt will desire sanlock as its locks manager.

Reply:
we can validate post-parsing or as a convertor, utilizing the following API:
`
PyObject* PyUnicode_AsASCIIString(PyObject *unicode)
    Return value: New reference.

    Encode a Unicode object using ASCII and return the result as Python bytes object. Error handling is “strict”. Return NULL if an exception was raised by the codec.
`

Reply:
we could use convert_to_pybytes shown above for both Py versions and not break anything on vdsm

Reply:
Supporting unicode and bytes is horrible hack, needed only for compatibillity with py 2. On py 3 code must use the correct type.

If we accept unicode, we should also return it in read_xxx, and we cannot do this because py 2 allowed anything, and there is no encoding.

Reply:
so this implies we'll need to straighten up the vdsm calls for lockspaces and resoruces once py3 sanlock is operational.

Reply:
anyway. i think the converter should be in a separate patch so it could be easily reverted once we drop py2 for good and then we can leave the "S" formatter for py3 as planned.